File: java/yass/test/ch/softappeal/yass/transport/socket/test/SslTest.java
package ch.softappeal.yass.transport.socket.test;

import ch.softappeal.yass.Exceptions;
import ch.softappeal.yass.InputStreamSupplier;
import ch.softappeal.yass.NamedThreadFactory;
import ch.softappeal.yass.remote.Server;
import ch.softappeal.yass.remote.test.ContractIdTest;
import ch.softappeal.yass.transport.socket.SimpleSocketTransport;
import ch.softappeal.yass.transport.socket.SocketBinder;
import ch.softappeal.yass.transport.socket.SocketConnector;
import ch.softappeal.yass.transport.socket.SslSetup;
import ch.softappeal.yass.transport.test.TransportTest;
import org.junit.Assert;
import org.junit.Test;

import javax.net.ServerSocketFactory;
import javax.net.SocketFactory;
import javax.net.ssl.SSLSocket;
import java.security.KeyStore;
import java.util.concurrent.Executors;

import static ch.softappeal.yass.transport.socket.test.SimpleSocketTransportTest.delayedShutdown;

public class SslTest extends TransportTest {

    private static void checkName(final String name) throws Exception {
        System.out.println("checkName");
        Assert.assertEquals(name, ((SSLSocket)SimpleSocketTransport.socket()).getSession().getPeerPrincipal().getName());
    }

    @SuppressWarnings("try")
    private static void test(final ServerSocketFactory serverSocketFactory, final SocketFactory socketFactory, final boolean needClientAuth) throws Exception {
        final var executor = Executors.newCachedThreadPool(new NamedThreadFactory("executor", Exceptions.STD_ERR));
        try (
            var closer = new SimpleSocketTransport(
                executor,
                MESSAGE_SERIALIZER,
                new Server(
                    ContractIdTest.ID.service(
                        new TestServiceImpl(),
                        (method, arguments, invocation) -> {
                            if (needClientAuth) {
                                checkName("CN=Client");
                            }
                            return invocation.proceed();
                        }
                    )
                )
            ).start(executor, SocketBinder.create(serverSocketFactory, SocketTransportTest.ADDRESS))
        ) {
            Assert.assertTrue(
                SimpleSocketTransport.client(MESSAGE_SERIALIZER, SocketConnector.create(socketFactory, SocketTransportTest.ADDRESS))
                    .proxy(
                        ContractIdTest.ID,
                        (method, arguments, invocation) -> {
                            checkName("CN=Server");
                            return invocation.proceed();
                        }
                    )
                    .divide(12, 4) == 3
            );
            System.out.println("ok");
        } finally {
            delayedShutdown(executor);
        }
    }

    private static final char[] PASSWORD = "StorePass".toCharArray();

    private static KeyStore readKeyStore(final String name) {
        return SslSetup.readKeyStore(InputStreamSupplier.create("../../certificates/" + name), PASSWORD);
    }

    private static final KeyStore SERVER_KEY = readKeyStore("Server.key.pkcs12");
    private static final KeyStore SERVER_CERT = readKeyStore("Server.cert.pkcs12");
    private static final KeyStore CLIENTCA_CERT = readKeyStore("ClientCA.cert.pkcs12");
    private static final KeyStore CLIENT_KEY = readKeyStore("Client.key.pkcs12");

    private static final String PROTOCOL = "TLSv1.2";
    private static final String CIPHER = "TLS_RSA_WITH_AES_128_CBC_SHA";

    @Test public void onlyServerAuthentication() throws Exception {
        test(
            new SslSetup(PROTOCOL, CIPHER, SERVER_KEY, PASSWORD, null).serverSocketFactory,
            new SslSetup(PROTOCOL, CIPHER, null, null, SERVER_CERT).socketFactory,
            false
        );
    }

    @Test public void clientAndServerAuthentication() throws Exception {
        test(
            new SslSetup(PROTOCOL, CIPHER, SERVER_KEY, PASSWORD, CLIENTCA_CERT).serverSocketFactory,
            new SslSetup(PROTOCOL, CIPHER, CLIENT_KEY, PASSWORD, SERVER_CERT).socketFactory,
            true
        );
    }

    @Test public void wrongServer() throws Exception {
        try {
            test(
                new SslSetup(PROTOCOL, CIPHER, SERVER_KEY, PASSWORD, null).serverSocketFactory,
                new SslSetup(PROTOCOL, CIPHER, null, null, CLIENTCA_CERT).socketFactory,
                false
            );
            Assert.fail();
        } catch (final Exception e) {
            e.printStackTrace();
        }
    }

    @Test public void wrongClientCA() throws Exception {
        try {
            test(
                new SslSetup(PROTOCOL, CIPHER, SERVER_KEY, PASSWORD, SERVER_CERT).serverSocketFactory,
                new SslSetup(PROTOCOL, CIPHER, CLIENT_KEY, PASSWORD, SERVER_CERT).socketFactory,
                true
            );
            Assert.fail();
        } catch (final Exception e) {
            e.printStackTrace();
        }
    }

}

File: kotlin/yass/test/ch/softappeal/yass/transport/socket/SslTest.kt
package ch.softappeal.yass.transport.socket

import ch.softappeal.yass.*
import ch.softappeal.yass.remote.*
import ch.softappeal.yass.remote.session.*
import ch.softappeal.yass.transport.*
import java.lang.reflect.*
import java.nio.file.*
import java.security.*
import javax.net.*
import javax.net.ssl.*
import kotlin.test.*

private fun test(
    serverSocketFactory: ServerSocketFactory, socketFactory: SocketFactory, needClientAuth: Boolean,
    uncaughtExceptionHandler: Thread.UncaughtExceptionHandler = Terminate
) {
    fun checkName(name: String) {
        println("checkName")
        assertEquals(name, (socket as SSLSocket).session.peerPrincipal.name)
    }
    useExecutor(uncaughtExceptionHandler) { executor, done ->
        socketServer(ServerSetup(Server(Service(calculatorId, CalculatorImpl, { _, _, invocation ->
            if (needClientAuth) checkName("CN=Client")
            invocation()
        })), messageSerializer), executor)
            .start(executor, socketBinder(address, serverSocketFactory)).use {
                useSyncClient(
                    socketClient(ClientSetup(messageSerializer), socketConnector(address, socketFactory))
                        .proxy(calculatorId, { _, _, invocation ->
                            checkName("CN=Server")
                            invocation()
                        })
                )
            }
        done()
    }
}

private fun failedTest(
    serverSocketFactory: ServerSocketFactory,
    socketFactory: SocketFactory,
    needClientAuth: Boolean
) = assertFailsWith<UndeclaredThrowableException> {
    test(
        serverSocketFactory, socketFactory, needClientAuth,
        Thread.UncaughtExceptionHandler { _, e ->
            e.printStackTrace()
        }
    )
}

private val PASSWORD = "StorePass".toCharArray()

private fun readKeyStore(name: String): KeyStore {
    return Files.newInputStream(Paths.get("../../certificates", name)).use { readKeyStore(it, PASSWORD) }
}

private val SERVER_KEY = readKeyStore("Server.key.pkcs12")
private val SERVER_CERT = readKeyStore("Server.cert.pkcs12")
private val CLIENTCA_CERT = readKeyStore("ClientCA.cert.pkcs12")
private val CLIENT_KEY = readKeyStore("Client.key.pkcs12")

private const val PROTOCOL = "TLSv1.2"
private val CIPHER_SUITES = listOf("TLS_RSA_WITH_AES_128_CBC_SHA", "TLS_RSA_WITH_AES_256_CBC_SHA")

class SslTest {
    @Test
    fun onlyServerAuthentication() {
        test(
            SslSetup(PROTOCOL, CIPHER_SUITES, SERVER_KEY, PASSWORD, null).serverSocketFactory,
            SslSetup(PROTOCOL, CIPHER_SUITES, null, null, SERVER_CERT).socketFactory,
            false
        )
    }

    @Test
    fun clientAndServerAuthentication() {
        test(
            SslSetup(PROTOCOL, CIPHER_SUITES, SERVER_KEY, PASSWORD, CLIENTCA_CERT).serverSocketFactory,
            SslSetup(PROTOCOL, CIPHER_SUITES, CLIENT_KEY, PASSWORD, SERVER_CERT).socketFactory,
            true
        )
    }

    @Test
    fun wrongServer() {
        failedTest(
            SslSetup(PROTOCOL, CIPHER_SUITES, SERVER_KEY, PASSWORD, null).serverSocketFactory,
            SslSetup(PROTOCOL, CIPHER_SUITES, null, null, CLIENTCA_CERT).socketFactory,
            false
        )
    }

    @Test
    fun wrongClientCA() {
        failedTest(
            SslSetup(PROTOCOL, CIPHER_SUITES, SERVER_KEY, PASSWORD, SERVER_CERT).serverSocketFactory,
            SslSetup(PROTOCOL, CIPHER_SUITES, CLIENT_KEY, PASSWORD, SERVER_CERT).socketFactory,
            true
        )
    }
}
